contracts/contracts/BLSExpanderDelegator.sol

```diff
@@ -0,0 +1,93 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity >=0.7.0 <0.9.0;
+pragma abicoder v2;
+
+import "./lib/VLQ.sol";
+import "./interfaces/IWallet.sol";
+import "./interfaces/IExpander.sol";
+
+// Aka gateway, but we only care about processBundle here
+interface IBundleProcessor {
+    function processBundle(
+        IWallet.Bundle memory bundle
+    ) external returns (
+        bool[] memory successes,
+        bytes[][] memory results
+    );
+}
+
+contract BLSExpanderDelegator {
+    uint8 constant BLS_KEY_LEN = 4;
+
+    IBundleProcessor gateway;
+    mapping(uint256 => IExpander) public expanders;
+
+    constructor(IBundleProcessor gatewayParam) {
+        gateway = gatewayParam;
+    }
+
+    function run(bytes calldata stream) external returns (
+        bool[] memory successes,
+        bytes[][] memory results
+    ) {
+        IWallet.Bundle memory bundle;
+
+        // The signature is just the last 64 bytes.
+        uint256 opsByteLen = stream.length - 64;
+        bundle.signature = abi.decode(stream[opsByteLen:], (uint256[2]));
+        stream = stream[:opsByteLen]; // Only keep what we still need
+
+        uint256 vlqValue;
+
+        // Get the number of operations upfront so that we can allocate the
+        // memory. This information is technically redundant but extracting it
+        // by decoding everything before we have a place to put it would add a
+        // lot of complexity. For <=127 operations it's only one extra byte.
+        // Otherwise 2 bytes.
+        (vlqValue, stream) = VLQ.decode(stream);
+
+        bundle.senderPublicKeys = new uint256[BLS_KEY_LEN][](vlqValue);
+        bundle.operations = new IWallet.Operation[](vlqValue);
+
+        uint256 opsDecoded = 0;
+
+        while (stream.length > 0) {
+            // First figure out which expander to use.
+            (vlqValue, stream) = VLQ.decode(stream);
```

Comment:
I would expect the logic in this index of expanders to be non-VLQ specific. Is there another way we could determine which expander is selected?

Reply:
I'm not sure what you mean. The idea is that the expander index is encoded as a VLQ. Alternatively, if we used a fixed width to encode which expander to use then we'd either have an uncomfortably finite number of slots for expanders or we'd be using more bytes than necessary.

Reply:
So for example, let's say I write a gzip expander. I would need to encode the expander index VLQ, but then the remaining operation/bundle data would be gzipped. I could see that being confusing/more complicated than using a non-VLQ format.

> ...have an uncomfortably finite number of slots for expanders or we'd be using more bytes than necessary

If an expander ever ran out of slots with a smaller index, you just deploy another expander delegator or nest one by having another delegator implement `IExpander`.

Reply:
I mean, it's VLQ, but for a particular expander it's always a fixed byte sequence. I don't really see the point of confusion there... you'll need to prepend a fixed byte sequence when using gzip/etc anyway.